File: katalyze/src/rules/Criteria/AllProblemsSolved.java
package rules.Criteria;

import model.Contest;
import model.Problem;
import model.Standings;
import rules.StandingsCriterion;

public class AllProblemsSolved implements StandingsCriterion {

	@Override
	public boolean isFulfilled(Standings standings) {
		Contest contest = standings.getContest();
		for (Problem problem : contest.getProblems()) {
			if (!standings.isSolved(problem)) {
				return false;
			}
		}
		return true;
	}

	@Override
	public String message() {
		return "All problems have now been solved";
	}
	

}

File: katalyze/src/main/kotlin/rules_kt/AllProblemsSolved.kt
package rules_kt

import kotlinx.coroutines.flow.flow
import model.Commentary
import model.EventImportance
import org.icpclive.cds.scoreboard.ContestStateWithScoreboard

class AllProblemsSolved : RuleInterface() {
    override val filters = listOf(FlowFilters::isFirstSolve, FlowFilters.byTrigger { contestInfo, runs ->
        runs.filter { it.isAccepted() }.map { it.problemId }.toSet()
            .containsAll(contestInfo.scoreboardProblems.map { it.id })
    })

    override suspend fun process(contestStateWithScoreboard: ContestStateWithScoreboard) = flow{
        emit(Commentary.fromRunUpdateState(
            contestStateWithScoreboard.state, EventImportance.Breaking
        ) { _, _ -> "All problems have now been solved" })
    }
}